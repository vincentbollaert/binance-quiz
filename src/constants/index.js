export * from './shapeState'
export * from './shapePropTypes'
export * from './mockData'
