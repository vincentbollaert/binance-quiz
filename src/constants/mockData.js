export const questions = [
  {
    id: 60,
    title: 'Enterprise Ethereum Alliance (EEA)',
    slug: 'ethereum-enterprise-alliance',
    excerpt: 'The industry’s first global standards organization…o accelerate the adoption of Enterprise Ethereum.'
  },
  {
    id: 56,
    title: 'Bitcoin Dominance',
    slug: 'bitcoin-dominance',
    excerpt: 'The ratio of Bitcoins market capitalization versu…e market capitalizations of all cryptocurrencies.'
  },
  {
    id: 68,
    title: 'Maker',
    slug: 'maker',
    excerpt: 'You become a “maker” when you place an order and i…its for someone else to fill/match with it later.'
  },
  {
    id: 167,
    title: 'Wei',
    slug: 'wei',
    excerpt: 'The smallest possible denomination of ether (ETH),…network. Often used when referring to gas prices.'
  },
  {
    id: 178,
    title: 'Tank',
    slug: 'tank',
    excerpt: 'A term adopted from the traditional financial mark…ative financial performance of a particular asset'
  },
  {
    id: 30,
    title: 'Encryption',
    slug: 'encryption',
    excerpt: 'Conversion of information or data into a secure co…t unauthorised access to the information or data.'
  },
  {
    id: 53,
    title: 'Binance Ecosystem Fund (BEF)',
    slug: 'binance-ecosystem-fund',
    excerpt: 'An initiative by Binance to collaborate with partn…re about the blockchain/cryptocurrency ecosystem.'
  },
  {
    id: 227,
    title: 'Linux',
    slug: 'linux',
    excerpt: 'A popular open-source operating system, created in…used in a wide range of devices around the world.'
  },
  {
    id: 255,
    title: 'Bitcoin Core',
    slug: 'bitcoin-core',
    excerpt: 'Leading implementation of the software that enable…e Bitcoin network. Initially released by Satoshi.'
  },
  {
    id: 270,
    title: 'Breakout',
    slug: 'breakout',
    excerpt: 'When the price of an asset moves outside of a defi…by breaking out of a support or resistance area. '
  },
  {
    id: 22,
    title: 'Ad Hoc',
    slug: 'ad-hoc',
    excerpt: 'A phrase of Latin origin that is used in modern En…an for this purpose specifically for this.'
  },
  {
    id: 137,
    title: 'Volatility',
    slug: 'volatility',
    excerpt: 'How quickly and how much the price of an asset cha…ual return of an asset over a set period of time.'
  },
  {
    id: 240,
    title: 'Candlestick',
    slug: 'candlestick',
    excerpt: 'A graph representation of price action that displa…se, high, and low points within a certain period.'
  },
  {
    id: 31,
    title: 'Eclipse Attack',
    slug: 'eclipse-attack',
    excerpt: 'When the majority of peers on the network are mali…des from receiving information from honest nodes.'
  },
  {
    id: 241,
    title: 'Sentiment',
    slug: 'sentiment',
    excerpt: 'The overall attitude of a community in regards to …hin investors towards a certain financial market.'
  },
  {
    id: 120,
    title: 'Pegged Currency',
    slug: 'pegged-currency',
    excerpt: 'A currency where the price is designed to remain t…egged to 1 USD. Also referred to as a stablecoin.'
  },
  {
    id: 177,
    title: 'Listing',
    slug: 'listing',
    excerpt: 'The addition of an asset to an exchange either as …oject team or as a decision made by the exchange.'
  },
  {
    id: 7,
    title: 'Fiat',
    slug: 'fiat',
    excerpt: 'Money that a government has declared to be legal tender.'
  },
  {
    id: 11,
    title: 'Phishing',
    slug: 'phishing',
    excerpt: 'A malicious attack where a bad actor will attempt …r to gain unauthorised access into their account.'
  },
  {
    id: 164,
    title: 'Commodity Futures Trading Commission (CFTC)',
    slug: 'commodity-futures-trading-commission',
    excerpt: 'US-based agency responsible for regulating the der…h includes options, swaps, and futures contracts.'
  },
  {
    id: 186,
    title: 'Break-Even Point (BEP)',
    slug: 'break-even-point',
    excerpt: 'The point where the total costs of an operation is equivalent to its current value or revenue.'
  },
  {
    id: 108,
    title: 'Masternode',
    slug: 'masternode',
    excerpt: 'Nodes on a network that often require a minimum am…n coin staked in order to access staking rewards.'
  },
  {
    id: 209,
    title: 'Peer-to-Peer (P2P)',
    slug: 'peer-to-peer',
    excerpt: 'When two or more computers are connected and share…esources without relying on a centralized server.'
  },
  {
    id: 253,
    title: 'Efficient Market Hypothesis (EMH)',
    slug: 'efficient-market-hypothesis-emh',
    excerpt: 'An economic theory stating financial markets refle…e information on the price of assets at any time.'
  },
  {
    id: 170,
    title: 'Maximum Supply',
    slug: 'maximum-supply',
    excerpt: 'Refers to the maximum number of coins or tokens th… will be ever created for a given cryptocurrency.'
  },
  {
    id: 193,
    title: 'Moon',
    slug: 'moon',
    excerpt: 'A colloquial expression used to describe a cryptoc…hat is experiencing a strong upward market trend.'
  },
  {
    id: 264,
    title: 'Bloom Filter',
    slug: 'bloom-filter',
    excerpt: 'A data structure that can be used to inform the user whether a particular item is part of a set'
  },
  {
    id: 157,
    title: 'Flippening',
    slug: 'flippening',
    excerpt: 'The name to describe the moment when/if Ethereum (…s Bitcoin (BTC) in terms of market capitalization'
  },
  {
    id: 27,
    title: 'All-Time High (ATH)',
    slug: 'all-time-high',
    excerpt: 'The highest price of a cryptocurrency in a quote currency, such as a dollar, BTC, ETH, or BNB'
  },
  {
    id: 168,
    title: 'Issuance',
    slug: 'issuance',
    excerpt: 'Generation of a new cryptocurrency which occurs in…epending on parameters specified by the creators.'
  },
  {
    id: 76,
    title: 'Web 1.0',
    slug: 'web-1',
    excerpt: 'The initial iteration of the web, when data was pr…d with hyperlinks. Also known as'
  },
  {
    id: 149,
    title: 'Delisting',
    slug: 'delisting',
    excerpt: 'The removal of an asset from an exchange either as…oject team or as a decision made by the exchange.'
  },
  {
    id: 207,
    title: 'One Cancels the Other Order (OCO)',
    slug: 'one-cancels-the-other-order',
    excerpt: 'A pair of orders that are created concurrently, but it is only possible for one of them to execute.'
  },
  {
    id: 217,
    title: 'Staking Pool',
    slug: 'staking-pool',
    excerpt: 'A pool where stakeholders combine their staking po…ir chance of successfully validating a new block.'
  },
  {
    id: 12,
    title: 'Plasma',
    slug: 'plasma',
    excerpt: 'An Ethereum off-chain scaling solution which may a…increase the transactions per second capablities.'
  },
  {
    id: 155,
    title: 'Confirmation Time',
    slug: 'confirmation-time',
    excerpt: 'The time elapsed when a transaction is submitted t…d the time it is recorded into a confirmed block.'
  },
  {
    id: 79,
    title: 'Breakeven Multiple',
    slug: 'breakeven-multiple',
    excerpt: 'The multiple of the current price by which an asse…ate in order to reach its previous all-time high.'
  },
  {
    id: 175,
    title: 'Wrapped Ether (WETH)',
    slug: 'wrapped-ether',
    excerpt: 'ERC-20 token representing Ether at a 1:1 ratio. It… ETH to ERC-20 tokens on decentralized platforms.'
  },
  {
    id: 43,
    title: 'Decentralized Application (DApp)',
    slug: 'decentralized-application',
    excerpt: 'Applications that run on a P2P network of computer…rnet without being controlled by a single entity.'
  },
  {
    id: 29,
    title: 'Daemon',
    slug: 'daemon',
    excerpt: 'A process operating in the background waiting for …ific event or condition in order to be activated.'
  },
  {
    id: 25,
    title: 'Algorithm',
    slug: 'algorithm',
    excerpt: 'A sequence of unambiguous instructions used for the purpose of solving a problem.'
  },
  {
    id: 72,
    title: 'Transactions Per second (TPS)',
    slug: 'transactions-per-second',
    excerpt: 'Refers to the amount of transactions that a blockchain is capable of processing each second.'
  },
  {
    id: 256,
    title: 'Isolated Margin',
    slug: 'isolated-margin',
    excerpt: 'The margin balance allocated to a position. Trader…estricting the amount allocated to each position.'
  },
  {
    id: 196,
    title: 'Block Reward',
    slug: 'block-reward',
    excerpt: 'The sum of coins awarded by the blockchain protoco… for each successfully mined and validated block.'
  },
  {
    id: 146,
    title: 'Benchmark',
    slug: 'benchmark',
    excerpt: 'A measurement standard which can be used to gauge …ce of a particular asset or investment portfolio.'
  },
  {
    id: 128,
    title: 'Block Height',
    slug: 'block-height',
    excerpt: 'The number of blocks in the chain between itself a…ck on that blockchain (genesis block or block 0).'
  },
  {
    id: 172,
    title: 'Sell Wall',
    slug: 'sell-wall',
    excerpt: 'A very large limit sell order or a cumulation of s…e same price level on an order book for an asset.'
  },
  {
    id: 169,
    title: 'Liquidity',
    slug: 'liquidity',
    excerpt: 'The ability to sell or buy any given asset without… fluctuations in the market price for that asset.'
  },
  {
    id: 174,
    title: 'Latency',
    slug: 'latency',
    excerpt: 'The time between submitting a transaction to a net… first confirmation of acceptance by the network.'
  },
  {
    id: 125,
    title: 'Resistance',
    slug: 'resistance',
    excerpt: 'A term in Technical Analysis (TA). When a price th…resistance. Usually compared with previous highs.'
  },
  {
    id: 75,
    title: 'Whiskers',
    slug: 'whiskers',
    excerpt: 'The lines extending from the colored bar in a cand…time frame. Also referred to as wicks or shadows.'
  },
  {
    id: 237,
    title: 'Deep Web',
    slug: 'deep-web',
    excerpt: 'The part of the web that is somehow hidden, i.e., … Google and other traditional web search engines.'
  },
  {
    id: 66,
    title: 'Know Your Customer (KYC)',
    slug: 'know-your-customer',
    excerpt: 'A standard procedure in the finance industry which…tify their customers and comply with KYC AML laws'
  },
  {
    id: 162,
    title: 'Mainnet',
    slug: 'mainnet',
    excerpt: 'A blockchain protocol which is fully developed and…ns are being broadcasted, verified, and recorded.'
  },
  {
    id: 73,
    title: 'User Interface (UI)',
    slug: 'user-interface',
    excerpt: 'The interface where interactions between humans an…tablishes how a user can interact with a machine.'
  },
  {
    id: 21,
    title: 'Absolute Advantage',
    slug: 'absolute-advantage',
    excerpt: 'An economics concept in which one party has a dire…ng a specific good or service over another party.'
  },
  {
    id: 163,
    title: 'Bid-Ask Spread',
    slug: 'bid-ask-spread',
    excerpt: 'The difference in price between the lowest asking … highest bid price on the order book for an asset'
  },
  {
    id: 141,
    title: 'Keccak',
    slug: 'keccak',
    excerpt: 'A versatile cryptographic function designed by Gui…n Daemen, Michaël Peeters, and Gilles Van Assche.'
  },
  {
    id: 171,
    title: 'Gas Limit',
    slug: 'gas-limit',
    excerpt: 'The maximum price a cryptocurrency user is willing…saction, or performing a smart contract function.'
  },
  {
    id: 39,
    title: 'Ask Price',
    slug: 'asking-price',
    excerpt: 'The lowest price a seller is willing to accept on … sell order when trading an asset on an exchange.'
  },
  {
    id: 37,
    title: 'Arbitrage',
    slug: 'arbitrage',
    excerpt: 'Buying and selling of assets over different market… advantage of differing prices on the same asset.'
  },
]

export const activeQuestion = {
  id: 68,
  title: 'Maker',
  slug: 'slug',
  excerpt: 'excerpt.',
  answers: ['Maker', 'Bitcoin Dominance', 'Encryption', 'Linux'],
  isFinalQuestion: false,
}
export const selectedAnswer = {
  id: 68,
  title: 'Maker',
  slug: 'maker',
  excerpt: 'You become a “maker”...',
  selectedAnswer: 'Maker',
  correctAnswer: 'Maker',
  timeToChoose: 10,
  isCorrect: true,
}

export const answerCorrect = {
  id: 1,
  selectedAnswer: 'answer',
  correctAnswer: 'answer',
  timeToChoose: 10,
  isCorrect: true,
}
export const answerIncorrect = {
  id: 2,
  selectedAnswer: 'answer incorrect',
  correctAnswer: 'answer',
  timeToChoose: 10,
  isCorrect: false,
}
