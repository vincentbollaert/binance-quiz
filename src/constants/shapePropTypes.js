import { bool, string } from 'prop-types'

export const PROP_ASYNC_STATUS = {
  isInitial: bool.isRequired,
  isBusy: bool.isRequired,
  isError: bool.isRequired,
  isSuccessful: bool.isRequired,
  errorMessage: string,
}

export const placeholder = ''
