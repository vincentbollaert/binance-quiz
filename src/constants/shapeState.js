export const IS_INITIAL = 'isInitial'
export const IS_BUSY = 'isBusy'
export const IS_ERROR = 'isError'
export const IS_SUCCESSFUL = 'isSuccessful'
export const ERROR_MESSAGE = 'errorMessage'

export const SHAPE_ASYNC_STATUS_INITIAL = {
  [IS_INITIAL]: true,
  [IS_BUSY]: false,
  [IS_ERROR]: false,
  [IS_SUCCESSFUL]: false,
  [ERROR_MESSAGE]: undefined,
}

export const SHAPE_ASYNC_STATUS_REQUESTED = {
  [IS_INITIAL]: false,
  [IS_BUSY]: true,
  [IS_ERROR]: false,
  [IS_SUCCESSFUL]: false,
  [ERROR_MESSAGE]: undefined,
}

export const SHAPE_ASYNC_STATUS_SUCCEEDED = {
  [IS_INITIAL]: false,
  [IS_BUSY]: false,
  [IS_ERROR]: false,
  [IS_SUCCESSFUL]: true,
  [ERROR_MESSAGE]: undefined,
}

export const SHAPE_ASYNC_STATUS_FAILED = (error) => ({
  [IS_INITIAL]: false,
  [IS_BUSY]: false,
  [IS_ERROR]: true,
  [IS_SUCCESSFUL]: false,
  [ERROR_MESSAGE]: error,
})
