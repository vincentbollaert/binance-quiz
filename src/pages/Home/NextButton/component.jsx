import React from 'react'
import { bool, shape, arrayOf, func } from 'prop-types'
import styled from 'styled-components'

import { FONT_SIZE_LG, FONT_SIZE_XLG, SELECTIVE_YELLOW } from '../../../styles'

const Wrap = styled.button`
  display: ${props => props.isShow ? 'flex' : 'none'};
  margin-left: auto;
  justify-content: center;
  background-color: #444;
  width: 11.4rem;
  height: 11.4rem;
  flex-shrink: 0;
  line-height: 1;
  text-transform: uppercase;
  font-size: ${FONT_SIZE_LG};
  cursor: pointer;
  text-transform: uppercase;
  font-weight: bold;
  pointer-events: none;
  color: #7e7e7e;

  ${props => props.isQuestionFinished && `
    pointer-events: all;
    background-color: ${SELECTIVE_YELLOW};
    color: #705603;

    &:hover {
      background-color: #ffba0a;
    };
  `};
`
const InnerWrap = styled.div`
  width: 4rem;
  height: 4rem;
  position: relative;
  justify-content: center;
  display: flex;
  align-items: center;
  font-size: ${FONT_SIZE_XLG};
`
const CurrentQuestion = styled.div`
  position: absolute;
  top: 0px;
  left: 4px;
`
const Divider = styled.div`
  width: 4.7rem;
  height: 1px;
  background: ${props => props.isQuestionFinished ? '#997e287a' : '#ffffff21'};
  position: absolute;
  margin-top: 0px;
  margin-left: 0px;
`
const NextQuestion = styled.div`
  position: absolute;
  right: 2px;
  bottom: 0px;
`

const NextButton = ({
  isQuizComplete,
  isActiveQuestionFinished,
  isLastQuestion,
  activeQuestion: { id, isFinalQuestion },
  selectedAnswers,
  onNextClick,
}) => {
  return (
    <Wrap
      isShow={!isQuizComplete}
      isQuestionFinished={isActiveQuestionFinished}
      onClick={() => onNextClick({ activeQuestionId: id, isFinalQuestion })}
    >
      {isLastQuestion ? 'Finish' : (
        <InnerWrap>
          <CurrentQuestion>{selectedAnswers.length + 1}</CurrentQuestion>
          <Divider isQuestionFinished={isActiveQuestionFinished} />
          <NextQuestion>10</NextQuestion>
        </InnerWrap>
      )}
    </Wrap>
  )
}

NextButton.propTypes = {
  isQuizComplete: bool.isRequired,
  isActiveQuestionFinished: bool.isRequired,
  isLastQuestion: bool.isRequired,
  activeQuestion: shape({}).isRequired,
  selectedAnswers: arrayOf(shape({})).isRequired,
  onNextClick: func.isRequired,
}

export default NextButton
