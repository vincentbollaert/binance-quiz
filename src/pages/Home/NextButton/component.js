import React, { memo } from 'react'
import { bool, shape, arrayOf, func } from 'prop-types'
import styled from 'styled-components'

import { FONT_SIZE_LG, FONT_SIZE_XLG, SELECTIVE_YELLOW, CHARLESTON_GREEN_DARKER } from '../../../styles'
import { SHAPE_QUIZ_QUESTION, SHAPE_SELECTED_ANSWER } from '../shapePropTypes'

const Wrap = styled.button`
  display: ${props => props.isShow ? 'flex' : 'none'};
  margin-left: auto;
  justify-content: center;
  background-color: ${CHARLESTON_GREEN_DARKER};
  width: 11.4rem;
  height: 11.4rem;
  flex-shrink: 0;
  line-height: 1;
  text-transform: uppercase;
  font-size: ${FONT_SIZE_LG};
  cursor: pointer;
  text-transform: uppercase;
  font-weight: bold;
  pointer-events: none;
  color: #7e7e7e;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);

  ${props => props.isQuestionFinished && `
    pointer-events: all;
    background-color: ${SELECTIVE_YELLOW};
    color: #705603;

    &:hover {
      background-color: #ffba0a;
    };
  `};
`
const InnerWrap = styled.div`
  width: 4rem;
  height: 4rem;
  position: relative;
  justify-content: center;
  display: flex;
  align-items: center;
  font-size: ${FONT_SIZE_XLG};
`
const CurrentQuestion = styled.div`
  position: absolute;
  top: 0px;
  left: 4px;
`
const Divider = styled.div`
  width: 4.7rem;
  height: 1px;
  background: ${props => props.isQuestionFinished ? '#997e287a' : '#ffffff21'};
  position: absolute;
  margin-top: 0px;
  margin-left: 0px;
`
const NextQuestion = styled.div`
  position: absolute;
  right: 2px;
  bottom: 0px;
`

const NextButton = ({
  isQuizComplete,
  activeQuestion: { id, isFinalQuestion },
  selectedAnswers,
  onNextClick,
}) => {
  const isActiveQuestionFinished = selectedAnswers.length && selectedAnswers[selectedAnswers.length - 1].id === id
  return (
    <Wrap
      isShow={!isQuizComplete}
      isQuestionFinished={isActiveQuestionFinished}
      onClick={() => onNextClick({ activeQuestionId: id, isFinalQuestion })}
      data-testid="component-next-button"
    >
      {isFinalQuestion ? 'Finish' : (
        <InnerWrap>
          <CurrentQuestion data-testid="current-question">{selectedAnswers.length + 1}</CurrentQuestion>
          <Divider isQuestionFinished={isActiveQuestionFinished} />
          <NextQuestion>10</NextQuestion>
        </InnerWrap>
      )}
    </Wrap>
  )
}

NextButton.propTypes = {
  isQuizComplete: bool.isRequired,
  activeQuestion: shape(SHAPE_QUIZ_QUESTION),
  selectedAnswers: arrayOf(shape(SHAPE_SELECTED_ANSWER)).isRequired,
  onNextClick: func.isRequired,
}

NextButton.defaultProps = {
  activeQuestion: {},
}

export default memo(NextButton)
