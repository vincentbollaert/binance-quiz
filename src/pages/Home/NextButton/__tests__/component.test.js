import '@testing-library/jest-dom'
import React from 'react'
import { render } from '@testing-library/react'
import { answerCorrect } from '../../../../constants'
import NextButton from '../component'

describe('NextButton', () => {
  const props = {
    isQuizComplete: false,
    activeQuestion: { id: 1, isFinalQuestion: false },
    selectedAnswers: [],
    onNextClick() {},
  }
  test('matches snapshot', () => {
    const { container } = render(<NextButton {...props} />)
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders elements', () => {
    const { queryByTestId, queryByText } = render(<NextButton {...props} isShow />)
    expect(queryByTestId('component-next-button')).toBeTruthy()
    expect(queryByText('Finish')).toBeFalsy()
    expect(queryByTestId('current-question')).toHaveTextContent('1')
  })
  test('renders correct activeQuestion number', () => {
    const { queryByTestId, rerender } = render(<NextButton {...props} selectedAnswers={[answerCorrect]} />)
    expect(queryByTestId('current-question')).toHaveTextContent('2')

    rerender(<NextButton {...props} selectedAnswers={[answerCorrect, answerCorrect]} />)
    expect(queryByTestId('current-question')).toHaveTextContent('3')
  })
  test('renders finish  if on last question', () => {
    const { queryByText, queryByTestId } = render(<NextButton {...props} activeQuestion={{ id: 1, isFinalQuestion: true }} />)

    expect(queryByTestId('current-question')).toBeFalsy()
    expect(queryByText('Finish')).toBeTruthy()
  })
})
