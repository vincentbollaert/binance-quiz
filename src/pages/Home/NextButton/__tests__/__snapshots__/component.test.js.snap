// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextButton matches snapshot 1`] = `
<button
  class="sc-AxjAm crptTO"
  data-testid="component-next-button"
>
  <div
    class="sc-AxirZ jqGPYj"
  >
    <div
      class="sc-AxiKw kogpRt"
      data-testid="current-question"
    >
      1
    </div>
    <div
      class="sc-AxhCb iNFifs"
    />
    <div
      class="sc-AxhUy lahxHt"
    >
      10
    </div>
  </div>
</button>
`;
